Add Openshift Metering role and playbook

chancez opened the pull request:
Add a role for installing the operator-metering stack for Openshift, and a playbook that uses this role.
Also update the common playbook and adds the openshift-metering playbook as a component.

Openshift metering is a specific configuration of https://github.com/operator-framework/operator-metering. The goal of this PR is to get it included into the ansible installer so we can begin testing in Openshift online, and so others have a consistent installation method.

Currently this expects openshift cluster monitoring to be installed, I wasn't quite sure the best way to have a dependency on that being installed and error if it's not, open to suggestions for handling that.

tongpu commented on playbooks/openshift-metering/README.md:
```diff
@@ -0,0 +1,32 @@
+
+## Installation
+
+To install the monitoring operator, set this variable:
```
This should probably be `s/monitoring/metering/`?

tongpu commented on roles/openshift_metering/README.md:
```diff
@@ -0,0 +1,49 @@
+
+## Installation
+
+To install the monitoring operator, set this variable:
```
Probably the same thing again `s/monitoring/metering/`

chancez commented:
Whoops, yeah, copied the monitoring role & playbook as a base and missed some stuff.

chancez commented:
/assign ironcladlou

ironcladlou commented on roles/openshift_metering/files/metering-helm-operator-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+      labels:
+        app: metering-helm-operator
+    spec:
+      securityContext:
```
This almost always causes trouble and probably needs removed; the platform will default it appropriately.

chancez replied:
This is well tested against openshift it shouldn't have issues unless it was set to false.

ironcladlou replied:
Tested in an Online cluster (e.g. free-int)? The GCP clusters you get from the CI tooling don't have the same security policy setup...

chancez replied:
That's...going to make validating things difficult. This PR is the pre-req for trying to deploy into online, so not there yet. Where can I get a copy of the that those settings so I can validate that this works?

chancez replied:
Confirmed in slack this shouldn't be an issue.

ironcladlou commented on roles/openshift_metering/files/metering-helm-operator-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+      - name: metering-helm-operator
+        image: "quay.io/coreos/chargeback-helm-operator:0.6.1-latest"
+        args: ["run-operator.sh"]
+        imagePullPolicy: Always
```
Would `IfNotPresent` work?

chancez replied:
It probably would, however this is what we've been using for a while without trouble. I prefer Always as you can be sure if an image is updated you're not going to see inconsistencies, but we only mutate non-release tags anyways.

ironcladlou commented on roles/openshift_metering/files/metering-helm-operator-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+          requests:
+            memory: "25Mi"
+            cpu: "50m"
+          limits:
```
Request defaults make sense, but do we need limits? With Prometheus, for example, we want dynamic growth to be possible for maximum portability across our various environments at different scales.

chancez replied:
I try to set both since quotas would do it if we don't, and it puts us into the guaranteed QOS. It shouldn't really need more I don't think. This is a deployment simple operator just doing templating using helm basically.

ironcladlou commented on roles/openshift_metering/files/metering-helm-operator-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+      - name: tiller
+        image: "quay.io/coreos/chargeback-helm-operator:0.6.1-latest"
+        args: ["/tiller"]
+        imagePullPolicy: Always
```
Is `IfNotPresent` ok?

ironcladlou commented on roles/openshift_metering/defaults/main.yml:
```diff
@@ -0,0 +1,6 @@
+---
+openshift_metering_operator_image: ''
+
+openshift_metering_namespace: openshift-metering
```
I think the general practice now is to be opinionated about this: exposing the namespace config is unnecessary (and even potentially harmful at this point)

chancez replied:
It's useful for testing since we can't install multiple copies into a single namespace yet. I want to be able to run this playbook against my cluster for testing so I'd like to be able to parameterize this.

ironcladlou replied:
No objection to making things configurable to facilitate development, but I would at least comment it with a "DO NOT CHANGE THIS" sort of thing or otherwise find a way to bury it. Any tips for this, @smarterclayton?

chancez replied:
I could add another boolean var as an additional check, like `openshift_metering_custom_namespace_enabled: false # ENABLE AT OWN RISK`

ewolinetz replied:
If you want to bury it, I'd recommend putting it into `vars/main.yml` instead

chancez replied:
Reading about it and it seems like this is probably a good improvement regardless. I made the change to moving it into `vars`. Is this sufficient?

ironcladlou commented on roles/openshift_metering/files/metering-helm-operator-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+          requests:
+            memory: "50Mi"
+            cpu: "50m"
+          limits:
```
As above re: limits

ironcladlou commented on roles/openshift_metering/files/metering-helm-operator-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+      restartPolicy: Always
+      terminationGracePeriodSeconds: 30
+      serviceAccount: metering-helm-operator
+      imagePullSecrets:
```
Unnecessary field

chancez replied:
Is it a big deal? These get rendered from some templates in the upstream repo and this is what's rendered when no imagePullSecrets are specified.

chancez replied:
Updated the upstream manifests to not output this, and pushed up the change.

ironcladlou commented on roles/openshift_metering/tasks/install.yaml:
```diff
@@ -0,0 +1,93 @@
+      resources: ["namespaces"]
+      verbs: ["get"]
+
+- name: Grant access to openshift-monitoring with namespace viewer role
```
s/monitoring/metering

ewolinetz commented on roles/openshift_metering/defaults/main.yml:
```diff
@@ -0,0 +1,6 @@
+---
+openshift_metering_operator_image: ''
```
For images, we've moved to a pattern that is based off the value of `l_os_registry_url`.
https://github.com/openshift/openshift-ansible/blob/master/roles/openshift_logging/defaults/main.yml#L14

Here is the commit that introduced that for reference: https://github.com/openshift/openshift-ansible/commit/a310ad860090d050b9c71a54ea0b1cc38f104db4

chancez replied:
That doesn't apply to this component yet, as we're still using images published on quay.io. We're still figuring out where the images will live, and how we publish for longer-term. This is still alpha quality.

ewolinetz commented on roles/openshift_metering/files/metering-helm-operator-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+        runAsNonRoot: true
+      containers:
+      - name: metering-helm-operator
+        image: "quay.io/coreos/chargeback-helm-operator:0.6.1-latest"
```
should this value be configurable?

chancez replied:
It is, it's replaced using yedit if `openshift_metering_operator_image` is set.

ewolinetz replied:
Is it not possible to provide this file as a jinja template instead? Then you could specify the value of the image as a `var` and it could default to the value above.

I know that with the service catalog there were times it wasn't possible to use a template since the files were actually sourced from another repo. Not sure if that is the same case here.

chancez replied:
It's possible, it's just more work than is needed since these files are just copied from another repo now, converting to Jinja means doing the conversion to Jinja anytime the upstream manifests change, so using yedit is a decent way to avoid that when we don't need to parameterize much.

ewolinetz replied:
Makes sense

ewolinetz commented on roles/openshift_metering/files/metering-helm-operator-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+            memory: "25Mi"
+            cpu: "50m"
+      - name: tiller
+        image: "quay.io/coreos/chargeback-helm-operator:0.6.1-latest"
```
Same question

chancez replied:
Same as above

ewolinetz commented on roles/openshift_metering/tasks/install.yaml:
```diff
@@ -0,0 +1,93 @@
+  with_fileglob:
+  - "*.yaml"
+
+- name: Copy admin client config
```
This should be unnecessary. Instead we should just reference it at `{{ openshift.common.config_base }}/master/admin.kubeconfig`

ewolinetz commented on roles/openshift_metering/tasks/install.yaml:
```diff
@@ -0,0 +1,93 @@
+    cp {{ openshift.common.config_base }}/master/admin.kubeconfig {{ mktemp.stdout }}/admin.kubeconfig
+  changed_when: false
+
+- name: Update Metering Helm Operator deployment image
```
Consider making the original yaml file a jinja template and using the `template` task instead here?

chancez replied:
Atm I'm trying to avoid unnecessary divergence from the upstream repo. The files are already templates (in a different form) in the upstream repo, and these are the rendered out results. Basically, the only things we need to make configurable for this are a few fields in the YAML which we can just replace. I asked @smarterclayton about how I should go about it and he recommended I went down this path.

ewolinetz commented on roles/openshift_metering/tasks/install.yaml:
```diff
@@ -0,0 +1,93 @@
+      value: "{{ openshift_metering_operator_image }}"
+  when: openshift_metering_operator_image != ""
+
+- name: Update Metering config
```
Same as above, re jinja template

chancez replied:
Same as above.

ewolinetz commented on roles/openshift_metering/tasks/install.yaml:
```diff
@@ -0,0 +1,93 @@
+      verbs: ["get"]
+
+- name: Grant access to openshift-metering with namespace viewer role
+  shell: >
```
Is there a reason why you're using shell and adding `|| true` to the end?

chancez replied:
The `oc_adm_policy_user` module doesn't allow specifying the name of the `ClusterRoleBinding` created, it re-uses the `ClusterRole`'s name. I want to be able to create a unique name so that if we were to run this playbook in a cluster with two different namespaces, the ClusterRoleBinding (which is global) it won't conflict with an existing one by using the unique name. The `|| true` is to handle the case where it already exists, in the case of re-running, then it should already be correctly set.

This made me realize I need to handle this ClusterRoleBinding uninstall however.

ewolinetz replied:
There is an option to update the `failed_when` condition on the task so that even if it already exists it won't fail out. 
In other places we've done a check prior to creating so that we only execute the task if it doesn't already exist.

I think one of those two would be preferable, in case the creation of the clusterrolebinding failed for another reason

chancez replied:
Gotcha, Ill try that.

ewolinetz commented on roles/openshift_metering/tasks/install.yaml:
```diff
@@ -0,0 +1,93 @@
+- name: Wait for the Report CRD to be created
+  command: >
+    {{ openshift_client_binary }} \
+      --config={{ mktemp.stdout }}/admin.kubeconfig \
```
update config to point to un-copied location

ewolinetz commented on roles/openshift_metering/tasks/uninstall.yaml:
```diff
@@ -0,0 +1,5 @@
+---
```
should we also be removing the CRD as part of removal/clean up?

chancez replied:
Good question. I think that's a bit more dangerous as there may be multiple installs of it and deleting the CRD would cause a cascading deletion of all instances. I think for now it's probably okay to leave around, and I believe this is what's happening in other playbooks.

ewolinetz commented on roles/openshift_metering/tasks/install.yaml:
```diff
@@ -0,0 +1,99 @@
+      --clusterrole \
+      openshift-metering-namespace-viewer \
+      --serviceaccount \
+      "{{ openshift_metering_namespace }}:metering" || true
```
do we still need `|| true` since we have the when condition now?

ewolinetz replied:
If not, can we also change this to be `command` instead of `shell`?

chancez replied:
Oh right, I forgot to update this part.

ewolinetz commented on roles/openshift_metering/vars/main.yml:
```diff
@@ -0,0 +1,2 @@
+---
+openshift_metering_namespace: openshift-metering
```
by no means a blocker, there is a pattern to prefix 'private' variables with "__" so this could be `__openshift_metering_namespace` which implies its private and shouldn't really be changed (but obviously you still can).

chancez replied:
That's a common pythonism but I grepped the codebase for this pattern and didn't find it in any yaml file.

ewolinetz replied:
Hmm, I know we do it in the logging roles. It may not be very wide spread anymore. It was a common practice before with our image names

https://github.com/openshift/openshift-ansible/blob/master/roles/openshift_logging_elasticsearch/vars/main.yml

chancez replied:
Heh, maybe I didn't give it enough time. I can do this.

ewolinetz commented:
one question regarding where shell is used, otherwise LGTM

ewolinetz commented:
/test install

chancez commented:
The jenkins failure looks to be this as far as I could determine:
```
+ sudo yum install -y openshift-ansible-3.10.1-9999.git.0.79527c7.el7
Loaded plugins: amazon-id, rhui-lb, search-disabled-repos
No package openshift-ansible-3.10.1-9999.git.0.79527c7.el7 available.
Error: Nothing to do
++ export status=FAILURE
++ status=FAILURE
+ set +o xtrace
```

vrutkovs commented:
/retest

chancez commented:
/test install

chancez commented:
Going to squash these commits.

chancez commented:
I forget where to find the CI logic, but it seems its trying 10 times to forward the credentials, getting `lost connection` each time, then it stops, but doesn't fail after the using up attempts, and then tries to SSH to the machine and fails at that point I think. Either way it seems like a flake, but I'd also love to fix the failure not happening in the correct spot: after being unable to SCP credentials passed the number of attempts.

chancez commented:
/retest

ewolinetz commented:
/test gcp

ewolinetz commented:
@chancez LGTM, if you want to squash to a single commit we can push this along (since its all part of the same deliverable)

chancez commented:
Squashed & rebased.

chancez commented:
/retest

ewolinetz commented:
/lgtm

ewolinetz commented:
/assign @vrutkovs

chancez commented:
@vrutkovs Any chance you could take a look at this this week?

vrutkovs commented:
Failed here with default settings and `openshift_metering_install: true` and `openshift_monitoring_deploy: true` set:
```
fatal: [vrutkovs_ac119853e5de-master-1]: FAILED! => {
    "changed": true, 
    "cmd": [
        "oc", 
        "--config=/etc/origin/master/admin.kubeconfig", 
        "--namespace", 
        "openshift-metering", 
        "apply", 
        "-f", 
        "/tmp/openshift-metering-ansible-AGljE4/metering-manifests"
    ], 
    "delta": "0:00:00.625843", 
    "end": "2018-06-26 10:19:03.314517", 
    "failed": true, 
    "rc": 1, 
    "start": "2018-06-26 10:19:02.688674"
}

STDOUT:

deployment.apps "metering-helm-operator" created
role.rbac.authorization.k8s.io "metering-helm-operator" created
rolebinding.rbac.authorization.k8s.io "metering-helm-operator" created
serviceaccount "metering-helm-operator" created


STDERR:

error: unable to recognize "/tmp/openshift-metering-ansible-AGljE4/metering-manifests/metering.yaml": no matches for kind "Metering" in version "chargeback.coreos.com/v1alpha1"


MSG:

non-zero return code
```

chancez commented:
Hmm, after seeing that I'm surprised I haven't noticed this sooner. CRD is not getting registered before the metering resource is being created. I'll just update it to create CRDs first.

ewolinetz commented:
/test install
looks like image building failed

vrutkovs commented on roles/openshift_metering/README.md:
```diff
@@ -0,0 +1,49 @@
+
+This role installs the OpenShift [Metering](https://github.com/operator-framework/operator-metering), which supports metering operators and applications in Kubernetes and producing reports on this metered information.
+
+### **NOTE: This component is unsupported in OCP at this time.**
```
Should ansible throw a warning when `deployment_type=='openshift-enterprise'`?

chancez replied:
I don't really know.

vrutkovs commented on roles/openshift_metering/tasks/install.yaml:
```diff
@@ -0,0 +1,116 @@
+      verbs: ["get"]
+
+- name: Check if namespace viewer clusterrolebinding exists
+  command: >
```
Use `oc_obj` in this playbook if possible

chancez replied:
Ah, I didn't realize this allowed me to create things, I'll use that, working on testing my change locally.

chancez commented:
> Does this deprecate Prometheus install?
No, this depends on the new openshift monitoring.

> Should this also install Grafana?
No, we don't support Grafana today, this is for a different use case than monitoring.

chancez commented:
/test gcp

chancez commented:
@vrutkovs I updated things to use oc_obj.

vrutkovs commented:
Fails with enabled metering:
```
TASK [openshift_metering : Wait for the Report CRD to be created] *****************************************************************************************************************************
task path: /code/openshift-ansible/roles/openshift_metering/tasks/install.yaml:99
...
FAILED - RETRYING: Wait for the Report CRD to be created (1 retries left).
...

STDERR:

No resources found.
Error from server (NotFound): customresourcedefinitions.apiextensions.k8s.io "reports.chargeback.coreos.com" not found
```

vrutkovs commented on roles/openshift_metering/tasks/install.yaml:
```diff
@@ -0,0 +1,114 @@
+          name: metering
+          namespace: "{{ __openshift_metering_namespace }}"
+
+- name: Install the Metering CRD
```
Nitpick: `oc_obj` here would be easier, see ASB:

```
- name: Create custom resource definitions for asb
  oc_obj:
    name: '{{ crd.metadata.name }}'
    kind: CustomResourceDefinition
    state: present
    content:
      path: /tmp/{{ crd.metadata.name }}
      data: '{{ crd }}'
  vars:
    crd: "{{ lookup('file', item) | from_yaml }}"
  with_fileglob:
    - 'files/*.automationbroker.io.yaml'
```
It would automatically wait for it to be created. It would also obsolete copying admin kubeconfig and copying templates to the host

chancez replied:
We copy files cause we mutate some of them, but I can definitely switch to oc_obj here too. I gotta use two since I have to ensure the Metering CRD is created before the Metering CR.

vrutkovs commented on roles/openshift_metering/tasks/install.yaml:
```diff
@@ -0,0 +1,114 @@
+  delay: 10
+  retries: 15
+
+- name: Install the Metering Helm Operator
```
ditto, `ob_obj` would simplify that

chancez commented:
Ah my cluster already has the CRDs and I'm not installing them here, so it was just lucky to work in my env. I'll fix that up and push some fixup commits.

chancez commented:
@vrutkovs It seems that switching to `oc_obj` has issues with running twice. It seems to call` oc create -f` but not `apply` and causes the following errors on my 2nd run:

```
TASK [openshift_metering : Create Metering CRDs] *******************************
Friday 29 June 2018  17:27:13 +0000 (0:00:01.960)       0:00:29.239 ***********
changed: [ci-chancez-4-ig-m-gl55] => (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/crds/report.crd.yaml)
changed: [ci-chancez-4-ig-m-gl55] => (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/crds/storagelocation.crd.yaml)
ok: [ci-chancez-4-ig-m-gl55] => (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/crds/metering.crd.yaml)
changed: [ci-chancez-4-ig-m-gl55] => (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/crds/reportgenerationquery.crd.yaml)
changed: [ci-chancez-4-ig-m-gl55] => (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/crds/scheduledreport.crd.yaml)
changed: [ci-chancez-4-ig-m-gl55] => (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/crds/reportprometheusquery.crd.yaml)
changed: [ci-chancez-4-ig-m-gl55] => (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/crds/prestotable.crd.yaml)
changed: [ci-chancez-4-ig-m-gl55] => (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/crds/reportdatasource.crd.yaml)

TASK [openshift_metering : Install Metering Helm Operator and Metering Config] ***
Friday 29 June 2018  17:27:28 +0000 (0:00:15.751)       0:00:44.991 ***********
failed: [ci-chancez-4-ig-m-gl55] (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/operator/metering-helm-operator-deployment.yaml) => {"changed": false, "item": "/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/operator/metering-helm-operator-deployment.yaml", "msg": {"cmd": "/bin/oc create -f /tmp/metering-helm-operator-xUYS51 -n openshift-metering", "results": {}, "returncode": 1, "stderr": "Error from server (AlreadyExists): error when creating \"/tmp/metering-helm-operator-xUYS51\": deployments.apps \"metering-helm-operator\" already exists\n", "stdout": ""}}
failed: [ci-chancez-4-ig-m-gl55] (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/operator/metering-helm-operator-role.yaml) => {"changed": false, "item": "/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/operator/metering-helm-operator-role.yaml", "msg": {"cmd": "/bin/oc create -f /tmp/metering-helm-operator-yfO7RB -n openshift-metering", "results": {}, "returncode": 1, "stderr": "Error from server (AlreadyExists): error when creating \"/tmp/metering-helm-operator-yfO7RB\": roles.rbac.authorization.k8s.io \"metering-helm-operator\" already exists\n", "stdout": ""}}
failed: [ci-chancez-4-ig-m-gl55] (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/operator/metering-helm-operator-rolebinding.yaml) => {"changed": false, "item": "/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/operator/metering-helm-operator-rolebinding.yaml", "msg": {"cmd": "/bin/oc create -f /tmp/metering-helm-operator-hJI5bg -n openshift-metering", "results": {}, "returncode": 1, "stderr": "Error from server (AlreadyExists): error when creating \"/tmp/metering-helm-operator-hJI5bg\": rolebindings.rbac.authorization.k8s.io \"metering-helm-operator\" already exists\n", "stdout": ""}}
failed: [ci-chancez-4-ig-m-gl55] (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/operator/metering.yaml) => {"changed": false, "item": "/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/operator/metering.yaml", "msg": {"cmd": "/bin/oc create -f /tmp/openshift-metering-YYj5sg -n openshift-metering", "results": {}, "returncode": 1, "stderr": "Error from server (AlreadyExists): error when creating \"/tmp/openshift-metering-YYj5sg\": meterings.chargeback.coreos.com \"openshift-metering\" already exists\n", "stdout": ""}}
failed: [ci-chancez-4-ig-m-gl55] (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/operator/metering-helm-operator-service-account.yaml) => {"changed": false, "item": "/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/operator/metering-helm-operator-service-account.yaml", "msg": {"cmd": "/bin/oc create -f /tmp/metering-helm-operator-IAuUv8 -n openshift-metering", "results": {}, "returncode": 1, "stderr": "Error from server (AlreadyExists): error when creating \"/tmp/metering-helm-operator-IAuUv8\": serviceaccounts \"metering-helm-operator\" already exists\n", "stdout": ""}}

PLAY RECAP *********************************************************************
ci-chancez-4-ig-m-gl55     : ok=52   changed=4    unreachable=0    failed=1
ci-chancez-4-ig-nb-c01d    : ok=1    changed=0    unreachable=0    failed=0
localhost                  : ok=15   changed=3    unreachable=0    failed=0


INSTALLER STATUS ***************************************************************
Initialization      : Complete (0:00:19)
Openshift Metering  : In Progress (0:00:29)
	This phase can be restarted by running: playbooks/openshift-metering/config.yml
Friday 29 June 2018  17:27:34 +0000 (0:00:05.804)       0:00:50.796 ***********
===============================================================================
openshift_metering : Create Metering CRDs ------------------------------ 15.75s
openshift_metering : Install Metering Helm Operator and Metering Config --- 5.80s
Gathering Facts --------------------------------------------------------- 2.77s
Gathering Facts --------------------------------------------------------- 2.26s
Run variable sanity checks ---------------------------------------------- 2.10s
openshift_metering : Create namespace viewer clusterrolebinding --------- 1.96s
Gather Cluster facts ---------------------------------------------------- 1.72s
Initialize openshift.node.sdn_mtu --------------------------------------- 1.64s
openshift_metering : Create openshift-metering project ------------------ 1.44s
get openshift_current_version ------------------------------------------- 1.12s
openshift_metering : Create metering namespace viewer cluster role ------ 1.04s
openshift_metering : Update Metering config ----------------------------- 0.90s
openshift_metering : Create temp directory for doing work in on target --- 0.78s
openshift_control_plane : slurp ----------------------------------------- 0.78s
Detecting Operating System from ostree_booted --------------------------- 0.77s
openshift_metering : Ensures /tmp/openshift-metering-ansible-XWMIHN/metering-manifests dir exists --- 0.77s
openshift_control_plane : stat ------------------------------------------ 0.66s
check for node already bootstrapped ------------------------------------- 0.64s
openshift_gcp : Add bootstrap node instances ---------------------------- 0.36s
openshift_sanitize_inventory : conditional_set_fact --------------------- 0.19s
```

The updated section relevant looks like this:
```
- name: Create Metering CRDs
  oc_obj:
    state: present
    kind: CustomResourceDefinition
    name: '{{ crd.metadata.name }}'
    namespace: "{{ __openshift_metering_namespace }}"
    content:
      path: /tmp/{{ crd.metadata.name }}
      data: '{{ crd }}'
  vars:
    crd: "{{ lookup('file', item) | from_yaml }}"
  with_fileglob:
  - 'files/crds/*.yaml'

- name: Install Metering Helm Operator and Metering Config
  oc_obj:
    state: present
    kind: CustomResourceDefinition
    name: '{{ crd.metadata.name }}'
    namespace: "{{ __openshift_metering_namespace }}"
    content:
      path: /tmp/{{ crd.metadata.name }}
      data: '{{ crd }}'
  vars:
    crd: "{{ lookup('file', item) | from_yaml }}"
  with_fileglob:
  - 'files/operator/*.yaml'

```

chancez commented:
After switching to `oc_obj` for some of these things I keep getting errors about the files not existing. See latest commit for changes.

chancez commented:
So I'm having some issues which I'm not really sure if its my setup or not. 

After more debugging, the later failures are actually due to the copy module failing to find the sources, which is even weirder. I tried from a previous commit, and it's actually showing the same stuff. 

Look for `Couldn\'t stat remote file: No such file or directory` in this gist: https://gist.github.com/chancez/fdd477027e5d179d86cbf0d2597ed461. 

Gonna keep debugging this and try to get it working.

vrutkovs commented on playbooks/openshift-metering/README.md:
````diff
@@ -0,0 +1,40 @@
+Then run:
+
+```bash
+ansible-playbook playbooks/openshift-metering/config.yml
````
This playbook is not created

vrutkovs commented on roles/openshift_metering/tasks/install.yaml:
```diff
@@ -0,0 +1,94 @@
+    kind: CustomResourceDefinition
+    name: "{{ obj.metadata.name }}"
+    namespace: "{{ __openshift_metering_namespace }}"
+    files:
```
`oc_obj` can't read straight from file, instead
```
    content:
      path: /tmp/{{ obj.metadata.name }}
      data: '{{ obj }}'
```
should be used

chancez replied:
I tried that and it made no difference.

chancez replied:
Ah, seems I was wrong, the message was just standard SFTP probing to see if the file exists. So I tried the above change, and I get the following:

```
TASK [openshift_metering : Install Metering manifests] *************************
task path: /usr/share/ansible/openshift-ansible/roles/openshift_metering/tasks/install.yaml:76
Monday 09 July 2018  18:01:02 +0000 (0:00:02.281)       0:00:48.741 ***********
 [WARNING]: Unable to find '/tmp/openshift-metering-ansible-dtCooq/metering-
manifests/operator' in expected paths (use -vvvvv to see paths)

```

chancez replied:
So I've made progress I tried using `files` for the operator manifests being put into a tempdir but it seems to be failing because lookup plugins run locally, not on the remote host. I'm reading about this and it seems like I'll have get the contents of each file and store them in a variable so I can avoid using `loopup('file', item)` and so I can avoid using `with_fileglob` since both run locally instead of remotely. I also tried `delegate_to: "{{ groups.oo_first_master.0 }}"` but I read lookup plugins don't obey delegation.

```
TASK [openshift_metering : Install Metering manifests] *************************
Monday 09 July 2018  18:24:27 +0000 (0:00:01.906)       0:00:50.520 ***********
 [WARNING]: Unable to find '/tmp/openshift-metering-ansible-tHiet7/metering-
manifests/operator' in expected paths (use -vvvvv to see paths)
changed: [ci-chancez-4-ig-m-gl55] => (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/crds/report.crd.yaml)
changed: [ci-chancez-4-ig-m-gl55] => (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/crds/storagelocation.crd.yaml)
ok: [ci-chancez-4-ig-m-gl55] => (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/crds/metering.crd.yaml)
changed: [ci-chancez-4-ig-m-gl55] => (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/crds/reportgenerationquery.crd.yaml)
changed: [ci-chancez-4-ig-m-gl55] => (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/crds/scheduledreport.crd.yaml)
changed: [ci-chancez-4-ig-m-gl55] => (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/crds/reportprometheusquery.crd.yaml)
changed: [ci-chancez-4-ig-m-gl55] => (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/crds/prestotable.crd.yaml)
changed: [ci-chancez-4-ig-m-gl55] => (item=/usr/share/ansible/openshift-ansible/roles/openshift_metering/files/crds/reportdatasource.crd.yaml)
```

vrutkovs commented on roles/openshift_metering/tasks/install.yaml:
```diff
@@ -0,0 +1,94 @@
+
+- name: Update Metering Helm Operator deployment image
+  yedit:
+    src: "{{ mktemp.stdout }}/metering-manifests/operator/metering-helm-operator-deployment.yaml"
```
This deployment YAML doesn't seem to be used anywhere, it should be created with `oc_obj` too:

```
- name: Apply Controller Deployment manifest
  oc_obj:
    state: present
    kind: Deployment
    name: "metering-helm-operator"
    namespace: "{{ __openshift_metering_namespace }}"
    files:
    - "{{ mktemp.stdout }}/metering-manifests/operator/metering-helm-operator-deployment.yaml"
  run_once: true
```

chancez commented:
After some messing around and learning more ansible I think I found something that works. The main issue was lookup plugins weren't working for the tempfiles, so I ended up using `fetch` and `register` to get access to these files for use with `oc_obj`.

chancez commented:
cc @vrutkovs this should be ready for another look. I can squash when things look good.

vrutkovs commented:
Looks good, please squash so I could put lgtm here

chancez commented:
@vrutkovs Squashed & rebased.

chancez commented:
/retest

chancez commented:
/test install

vrutkovs commented:
/lgtm

Thanks!
